### ISSUE TITLE
Soundstream loss doesn't decrease after 1167 steps - version 0.7.1

### ISSUE THREAD

**Issue by yigityu** [lucidrains/audiolm-pytorch#61]
Hi,

First of all, thank you for this project and all the other open source projects you're doing. I'm a big fan of your work.

I was training with the latest version on LibriSpeech dataset and looks like recon_loss shoots up and training goes to nowhere afterwards. I didn't seem to have this with previous releases, but I will roll back, try again and report results here, but this might be a regression with latest changes? I wanted to post it if it would help anyone.

<img width="490" alt="image" src="https://user-images.githubusercontent.com/35917180/214087476-c4ca36e6-657b-4f43-b7de-5041647fa595.png">

**Comment by lucidrains** [comment]
@yigityu hey, thanks for reporting

could you try turning [this setting off](https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/soundstream.py#L317) ? if that doesn't work, also turn off [this](https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/soundstream.py#L316)

**Comment by yigityu** [comment]
Trying now.

**Comment by lucidrains** [comment]
@yigityu how high is your learning rate? after adding the local attentions, learning rate may need to be lowered to around `2e-4`

**Comment by yigityu** [comment]
I kept it at default of 3e-4 until now. I'll then do two trials - one to confirm that turning these options off works after 1000+ steps (running now), and running with these options + lowered learning rate. Will report results.

**Comment by lucidrains** [comment]
@yigityu thank you! :pray:

**Comment by djqualia** [comment]
FWIW 0.7.1 doesn't appear to be training too well for me either.  I'm 1200 steps in (~1 day) and the samples are just high pitched noise.  With 0.5.1 at the same # of steps the samples were starting to resemble the source.  I'm going to let 0.7.1 train a while longer and will try some of the suggestions here...

**Comment by yigityu** [comment]
I'm afraid turning off these two options didn't change the result.

Tuning down the lr to 2e-4 just delayed the problem to 5000 steps. Here are the graphs for tuned down lr:

<img width="490" alt="image" src="https://user-images.githubusercontent.com/35917180/214231972-59206e26-bdee-47f2-a9ac-7881454d47a7.png">

**Comment by lucidrains** [comment]
that only leaves this change https://github.com/lucidrains/audiolm-pytorch/commit/36c3954033caa535c397574c891da5711f056a37 as the culprit, between 0.5.1 and 0.7, I'll switch it back soon

**Comment by lucidrains** [comment]
thank you for these experiments and your patience

**Comment by yigityu** [comment]
Meanwhile I rolled back to daeedb2 and it's still stable after 9000 steps.

**Comment by yigityu** [comment]
> thank you for these experiments and your patience

my pleasure.

**Comment by lucidrains** [comment]
@yigityu ok, do you want to try [0.7.2](https://github.com/lucidrains/audiolm-pytorch/commit/e6eaabdb6691780afe7e6d74f46341361cdb621e) ?

**Comment by BlackFox1197** [comment]
@yigityu what were your training parameters? 
In 0.5.1 after 200,000 steps (~3 days) I still only had pitched noise, the last release which started to resemble the signal (after 32,000 steps) was 0.2.3. But I`m beginning to think I do not use the framework correctly as it seems to work with others.

**Comment by yigityu** [comment]
> @yigityu ok, do you want to try [0.7.2](https://github.com/lucidrains/audiolm-pytorch/commit/e6eaabdb6691780afe7e6d74f46341361cdb621e) ?

Sure, I will try out and let you know the results.

> @yigityu what were your training parameters? In 0.5.1 after 200,000 steps (~3 days) I still only had pitched noise, the last release which started to resemble the signal (after 32,000 steps) was 0.2.3. But I`m beginning to think I do not use the framework correctly as it seems to work with others.

I'm using default params listed in readme, but I didn't really yet achieve anything but noise, I'm just trying to stabilize training for at least couple of days but I'm really hopeful that we'll get it working.

**Comment by BlackFox1197** [comment]
@yigityu have a look at my trials: https://github.com/lucidrains/audiolm-pytorch/issues/57#issuecomment-1401986720
If you have the time you could also try it with 0.2.3. I would really be interested if this works better for you too

**Comment by yigityu** [comment]
Both runs in progress,

Run with 0.7.2 - with lr=3e-4 same issue at around 16000 steps. Trying now with a smaller learning rate from an earlier checkpoint.
Run with daeedb2 with lr=3e-4 losses didn't seem to stabilize enough. Reduced lr to 2e-4 and now at 60k steps. Still noise but it sounds somewhat less garbage.

@BlackFox1197 I think we're probably getting similar results here. I will keep it running little bit further, but might need to go back to 0.2.3 to see if it seems better.

**Comment by lucidrains** [comment]
yeah I think it is fine now, just keep training as long as nothing explodes. djqualia said he trained to 10 days on A6000, for reference

**Comment by djqualia** [comment]
Full disclosure, that 10 day training was at version 0.2.x or below... please see the other related thread where I think something has regressed since 0.2.3 which is the best results I've had.  I wish it was easy to test these different newer versions in parallel to pinpoint but it's rather time consuming.  I'll continue to try and collect better A/B data and share here....

**Comment by lucidrains** [comment]
@djqualia oh noes

how sure are you that it regressed? the only other offending diff would be https://github.com/lucidrains/audiolm-pytorch/commit/8259b0d03ce0fe49d9dcc49ae29e0ccbb704e7bc maybe i should roll that back

**Comment by lucidrains** [comment]
the other thing is that anything prior to 0.3 had [a bug](https://github.com/lucidrains/audiolm-pytorch/issues/48) for the discriminators, so i'm not sure how it could have learned in the past

**Comment by lucidrains** [comment]
@yigityu @BlackFox1197 @djqualia ok, decided to just bring back the complex stft discriminator in [0.7.3](https://github.com/lucidrains/audiolm-pytorch/commit/2ebfd1c18609c839ad74ff4fe279b2937aeac248) could one of you run this and see if this is any better

if it does not, we'll just keep rolling back changes until it works again

**Comment by lucidrains** [comment]
i'll set aside some time to run experiments too, maybe this Sunay

**Comment by djqualia** [comment]
Sure thing, I started a new training run with 0.7.3 just now, will report back tomorrow...

**Comment by lucidrains** [comment]
@djqualia thank you!

**Comment by djqualia** [comment]
Just a quick check before I go to work (too soon to tell I think), but 0.7.3 appears to be a bit better.  I'm only on step 720 with my config, but I'm hearing something resembling the source sounds behind some high pitch noise.  Hopefully it keeps improving!

I'm also running a separate training on 0.7.2 (not apples to apples since it's separate hardware), to try and determine if the regression was fixed with that too (just in case)

**Comment by lucidrains** [comment]
@djqualia tremendous! keep us updated 🙏

**Comment by BlackFox1197** [comment]
I have 8000 steps now in 0.7.3 and its way better than 0.7.1 or 0.5.1
![7_3_8000](https://user-images.githubusercontent.com/33451580/214969418-702f3f60-d579-4ffe-a54c-583eaa52a5a0.png)
I'm gonna try to hit 200,000 - 300,000 steps over the weekend, but it looks very good!

**Comment by lucidrains** [comment]
@BlackFox1197 🥳 🚀  hurray! time to democratize audio encoding 😄 

however, this would also mean [this issue](https://github.com/lucidrains/audiolm-pytorch/issues/31) is back to being unresolved 😢

**Comment by lucidrains** [comment]
i'll play around with [these settings](https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/soundstream.py#L245) over the weekend

i got them from the stft discriminator from encodec, but perhaps something isn't done right

**Comment by lucidrains** [comment]
the encodec non-complex stft discriminator is also using weight normalization while I'm using gradient penalty

could be another reason why it isn't working

**Comment by lucidrains** [comment]
what would actually be best is if the pytorch team resolves their issue

**Comment by djqualia** [comment]
yep, i agree and can also confirm that 0.7.3 is doing better!

my comparison to 0.7.2 is less conclusive/scientific, but 0.7.2 does seem worse to me.  i wonder if that means the change from 0.7.1->0.7.2 was unnecessary?

**Comment by lucidrains** [comment]
@djqualia woohoo!

i'm not sure re: 0.7.1 -> 0.7.2 @yigityu seems to have thought it became more stable, but we will probably have to just try the encodec activation configuration on the complex stft discriminator (and compare to the way it is now + complex stft discr)

**Comment by lucidrains** [comment]
I'm sort of in the "if it ain't broke don't try to fix it" stage 😂

**Comment by BlackFox1197** [comment]
ok unfortunately I have bad news. 
It looked fine at first, but After 18,000 steps this is what I get for an error:
`soundstream total loss: nan, soundstream recon loss: nan | discr (scale 1) loss: nan | discr (scale 0.5) loss: nan | discr (scale 0.25) loss: nan`

and it started exploding after 16,000 steps
In addition, after 12,000 steps it really degraded, 
![7_3_12000_bad](https://user-images.githubusercontent.com/33451580/215041226-27d4f24b-4532-46fe-b828-d4f099198160.png)
And with the checkpoint at 18,000 steps, it does not event produce an output anymore....

**Comment by lucidrains** [comment]
@BlackFox1197 ahh yeah, so that is called divergence. could you get your loss curves and share those?

Try lowering the learning rate, or removing attention in the next run

**Comment by lucidrains** [comment]
that's fine though, it is basically working early on, not that concerned

**Comment by lucidrains** [comment]
@BlackFox1197 the typical thing we do when adversarial training is involved is to save checkpoints and pick the one just before it explodes

It still shouldn't explode this early, but just so you know

**Comment by lucidrains** [comment]
with the new https://google-research.github.io/seanet/musiclm/examples/, should probably just get my hands dirty and start training

basically same as what is in the repo

**Comment by lucidrains** [comment]
oh, they actually have a new audio clip model in here called MuLan... ok, gotta build that here then too (or separate repo)

Regardless, getting soundstream trained is a priorty now

**Comment by yigityu** [comment]
I'm at 25k steps with version 0.7.3 and seems good so far - still a lot of noise but stable training. 

Looks like new challenges are piling up @lucidrains, let's get this soundstream trained :)

**Comment by lucidrains** [comment]
@yigityu yesss lets do this! stable diffusion moment for audio (sounds, speech, music, whatever) here we come!

**Comment by lucidrains** [comment]
soundstream is the linchpin

**Comment by lucidrains** [comment]
@yigityu is your training run on 0.7.3 with all the settings turned on? (attention and learned EMA?)

**Comment by yigityu** [comment]
Yes, those are all turned on.

**Comment by lucidrains** [comment]
@yigityu learning rate still at `2e-4` ? @BlackFox1197 how big is your training set? are you doing LibreSpeech?

**Comment by yigityu** [comment]
I'm keeping that 3e-4, but was planning to reduce it depending on how it goes.

**Comment by djqualia** [comment]
FWIW my training at 0.7.3 is still stable ~1.5 days in.  I am using a learning rate of 3e-4 and effective batch size of 96 (in my experience, with lower batch sizes you might need to lower the learning rate), with everything else as it defaults in 0.7.3.

+1 to the seizing the "stable diffusion moment" for audio, and the importance of Soundstream to enable all that here...  thanks @lucidrains for your work!

**Comment by lucidrains** [comment]
@djqualia no problem! happy training everyone!

**Comment by yigityu** [comment]
I know this issue is closed - but I have some good news so wanted to give an update on my runs if it helps anyone:

My run with daeedb2 with lr reduced to 2e-4 seems stable now. 155k steps and I can hear a human voice in the background :)

**Comment by BlackFox1197** [comment]
Im around 55000 steps and it seems stable too after reducing the learning rate. :) I can also remotely hear a voice already :)
